onmt/bin/train.py

```diff
@@ -24,38 +26,85 @@
 
 def prepare_fields_transforms(opt):
     """Prepare or dump fields & transforms before training."""
-    DynamicArgumentParser.validate_prepare_opts(opt)
-
     transforms_cls = get_transforms_cls(opt._all_transform)
     specials = get_specials(opt, transforms_cls)
 
     fields = build_dynamic_fields(
         opt, src_specials=specials['src'], tgt_specials=specials['tgt'])
-    save_fields(opt, fields)
 
-    # prepare pretrained embeddings, if any
+    # maybe prepare pretrained embeddings, if any
     prepare_pretrained_embeddings(opt, fields)
 
-    transforms = make_transforms(opt, transforms_cls, fields)
-    save_transforms(opt, transforms)
+    if opt.dump_fields:
+        save_fields(fields, opt.save_data, overwrite=opt.overwrite)
+    if opt.dump_transforms or opt.n_sample != 0:
+        transforms = make_transforms(opt, transforms_cls, fields)
+    if opt.dump_transforms:
+        save_transforms(transforms, opt.save_data, overwrite=opt.overwrite)
     if opt.n_sample != 0:
         logger.warning(
             "`-n_sample` != 0: Training will not be started. "
             f"Stop after saving {opt.n_sample} samples/corpus.")
         save_transformed_sample(opt, transforms, n_sample=opt.n_sample)
-        sys.exit("Sample saved, please check it before restart training.")
+        logger.info(
+            "Sample saved, please check it before restart training.")
+        sys.exit()
+    return fields, transforms_cls
+
+
+def _init_train(opt):
+    """Common initilization stuff for all training replica."""
+    DynamicArgumentParser.validate_prepare_opts(opt)
+
+    if opt.train_from:
+        # Load checkpoint if we resume from a previous training.
+        checkpoint = load_checkpoint(ckpt_path=opt.train_from)
+        fields = load_fields(opt.save_data, checkpoint)
+        transforms_cls = get_transforms_cls(opt._all_transform)
+        if (hasattr(checkpoint["opt"], '_all_transform') and
```

Comment:
This check is a bit strange, since the line above assumes it exists anyways.

Reply:
It's to avoid failing when `-train_from` an old model (which doesn't have `_all_transforms` in the checkpoint's opt).